dbms/src/Interpreters/Join.h

```diff
@@ -271,6 +324,60 @@ class Join
     // only use for left semi joins.
     const String match_helper_name;
 
+    struct BuildPartition
+    {
+        BlocksList blocks;
+        Blocks original_blocks;
+        size_t rows{0};
+        size_t bytes{0};
+    };
+
+    struct ProbePartition
+    {
+        Blocks blocks;
+        size_t rows{0};
+        size_t bytes{0};
+    };
+
+    struct JoinPartition
+    {
+        /// mutex to protect concurrent modify partition
+        /// note if you wants to acquire both build_probe_mutex and partition_mutex,
+        /// please lock build_probe_mutex first
+        std::mutex partition_mutex;
+        BuildPartition build_partition;
+        ProbePartition probe_partition;
+        ArenaPtr pool;
+        bool spill{};
+        /// only update this field when spill is enabled. todo support this field in non-spill mode
+        std::atomic<size_t> memory_usage{0};
+
+        void insertBlockForBuild(Block && block)
+        {
+            size_t rows = block.rows();
+            size_t bytes = block.bytes();
+            build_partition.rows += rows;
+            build_partition.bytes += bytes;
+            build_partition.blocks.push_back(block);
+            build_partition.original_blocks.push_back(std::move(block));
+            memory_usage += bytes;
+        }
+
+        void insertBlockForProbe(Block && block)
+        {
+            size_t rows = block.rows();
+            size_t bytes = block.bytes();
+            probe_partition.rows += rows;
+            probe_partition.bytes += bytes;
+            probe_partition.blocks.push_back(std::move(block));
```

Comment:
Although not used in probe stage, but I think we should update the `memory_usage` here?

Reply:
> Although not used in probe stage, but I think we should update the `memory_usage` here?

done.